nistats/tests/test_glm_reporter/glm_reporter_examples/first_level_models/report_fiac_analysis.py

```diff
@@ -0,0 +1,102 @@
+import os
+
+import numpy as np
+import pandas as pd
+
+from nilearn.image import mean_img
+
+from nistats import datasets
+from nistats._utils.testing import make_reporter_example_result_path
+from nistats.first_level_model import FirstLevelModel
+from nistats.reporting import generate_subject_stats_report
+
+def prep_data():
+    write_dir = os.path.join(os.getcwd(), 'results')
+    if not os.path.exists(write_dir):
+        os.mkdir(write_dir)
+
+    data = datasets.fetch_fiac_first_level()
+    fmri_img = [data['func1'], data['func2']]
+
+    # Create a mean image for plotting purpose
+    mean_img_ = mean_img(fmri_img[0])
+    design_files = [data['design_matrix1'], data['design_matrix2']]
+
+    design_matrices = [pd.DataFrame(np.load(df)['X']) for df in design_files]
+    n_columns = design_matrices[0].shape[1]
+    return fmri_img, data, design_matrices, n_columns, write_dir
+
+# GLM estimation
+
+
+# Compute fixed effects of the two runs and compute related images
+# For this, we first define the contrasts as we would do for a single session
+
+def pad_vector(contrast_, n_columns):
+    """A small routine to append zeros in contrast vectors"""
+    return np.hstack((contrast_, np.zeros(n_columns - len(contrast_))))
+
+def compute_statistics(fmri_img, data, design_matrices, n_columns, write_dir):
+    fmri_glm = FirstLevelModel(mask=data['mask'], minimize_memory=True)
+    fmri_glm = fmri_glm.fit(fmri_img, design_matrices=design_matrices)
+    contrasts = {'SStSSp_minus_DStDSp': pad_vector([1, 0, 0, -1], n_columns),
+                 'DStDSp_minus_SStSSp': pad_vector([-1, 0, 0, 1], n_columns),
+                 'DSt_minus_SSt': pad_vector([-1, -1, 1, 1], n_columns),
+                 'DSp_minus_SSp': pad_vector([-1, 1, -1, 1], n_columns),
+                 'DSt_minus_SSt_for_DSp': pad_vector([0, -1, 0, 1], n_columns),
+                 'DSp_minus_SSp_for_DSt': pad_vector([0, 0, -1, 1], n_columns),
+                 'Deactivation': pad_vector([-1, -1, -1, -1, 4], n_columns),
+                 'Effects_of_interest': np.eye(n_columns)[:5]}
+    
+    for index, (contrast_id, contrast_val) in enumerate(contrasts.items()):
+    
+        z_map = fmri_glm.compute_contrast(
+            contrast_val, output_type='z_score')
+    
+        # Write the resulting stat images to file
+        z_image_path = os.path.join(write_dir, '%s_z_map.nii.gz' % contrast_id)
+        z_map.to_filename(z_image_path)
+
+    # Comparing session-specific and fixed effects.
+    
+    contrast_id = 'Effects_of_interest'
```

Comment:
I was expecting this function to return all the contrasts defined in `contrasts`.  The full script does actually become simpler: 

```{python}
import os

import numpy as np
import pandas as pd

from nilearn.image import mean_img

from nistats import datasets
from nistats._utils.testing import make_reporter_example_result_path
from nistats.first_level_model import FirstLevelModel
from nistats.reporting import generate_subject_stats_report

def prep_data():
    write_dir = os.path.join(os.getcwd(), 'results')
    if not os.path.exists(write_dir):
        os.mkdir(write_dir)

    data = datasets.fetch_fiac_first_level()
    fmri_img = [data['func1'], data['func2']]

    # Create a mean image for plotting purpose
    mean_img_ = mean_img(fmri_img[0])
    design_files = [data['design_matrix1'], data['design_matrix2']]

    design_matrices = [pd.DataFrame(np.load(df)['X']) for df in design_files]
    n_columns = design_matrices[0].shape[1]
    return fmri_img, data, design_matrices, n_columns, write_dir


def pad_vector(contrast_, n_columns):
    """A small routine to append zeros in contrast vectors"""
    return np.hstack((contrast_, np.zeros(n_columns - len(contrast_))))


def compute_statistics(fmri_img, data, design_matrices, n_columns, write_dir):
    fmri_glm = FirstLevelModel(mask=data['mask'], minimize_memory=True)
    fmri_glm = fmri_glm.fit(fmri_img, design_matrices=design_matrices)
    contrasts = {'SStSSp_minus_DStDSp': pad_vector([1, 0, 0, -1], n_columns),
                 'DStDSp_minus_SStSSp': pad_vector([-1, 0, 0, 1], n_columns),
                 'DSt_minus_SSt': pad_vector([-1, -1, 1, 1], n_columns),
                 'DSp_minus_SSp': pad_vector([-1, 1, -1, 1], n_columns),
                 'DSt_minus_SSt_for_DSp': pad_vector([0, -1, 0, 1], n_columns),
                 'DSp_minus_SSp_for_DSt': pad_vector([0, 0, -1, 1], n_columns),
                 'Deactivation': pad_vector([-1, -1, -1, -1, 4], n_columns),
                 'Effects_of_interest': np.eye(n_columns)[:5]}


    z_maps = {}
    contrasts_id = []
    for (contrast_id, contrast_val) in contrasts.items():
        contrasts_id.append(contrast_id)
        z_map = fmri_glm.compute_contrast(
            contrast_val, output_type='z_score')
        # Write the resulting stat images to file
        z_image_path = os.path.join(write_dir, '%s_z_map.nii.gz' % contrast_id)
        z_map.to_filename(z_image_path)
        z_maps[contrast_id] = z_image_path

    return contrasts, z_maps, fmri_glm, design_matrices
    

def create_report_fiac_analysis():
    fmri_img, data, design_matrices, n_columns, write_dir = prep_data()
    (contrasts, z_maps, fmri_glm, design_matrices) = compute_statistics(
        fmri_img, data, design_matrices, n_columns, write_dir)

    output_filepath = make_reporter_example_result_path(__file__)
    
    generate_subject_stats_report(
            stats_report_filename=output_filepath,
            contrasts=contrasts,
            z_maps=z_maps,
            mask=fmri_glm.masker_.mask_img_,
            design_matrices=design_matrices,
            anat=mean_img(fmri_img[0]),
            threshold=3.0,
            )


if __name__ == '__main__':
    create_report_fiac_analysis()
```